File: src/year2015/day04/part1.cc
#include <cstdint>
#include <format>
#include <string>

#include "../../helpers/md5.h"
#include "../solutions.h"

auto Year2015::Day04::Part1(const std::string& input) -> std::string {
  for (uint64_t i = 0;; i++) {
    Helpers::MD5 hasher;
    const auto str = std::format("{}{}", input, i);
    hasher.consume(std::vector<uint8_t>(str.begin(), str.end()));
    const auto hash = hasher.compute();
    if (hash[0] == 0x00 && hash[1] == 0x00 && (hash[2] & 0xF0) == 0x00) {
      return std::to_string(i);
    }
  }
}

File: src/year2015/day04/part1.rs
use crate::helpers::md5::MD5;

pub fn part1(input: String) -> Result<String, String> {
  let mut i: u64 = 0;
  loop {
    let mut hasher = MD5::new();
    hasher.consume(input.clone().into());
    hasher.consume(i.to_string().into());
    let hash = hasher.compute();
    if hash[0] == 0x00 && hash[1] == 0x00 && (hash[2] & 0xF0) == 0x00 {
      return Ok(i.to_string());
    }
    if i == u64::MAX {
      break;
    }
    i += 1;
  }
  return Err("Exhausted all 64 bit unsigned integers while mining block".to_string());
}
